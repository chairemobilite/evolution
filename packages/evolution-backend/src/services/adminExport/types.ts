/*
 * Copyright 2024, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

type ExportOptions = {
    /** Specifies which set of responses should be exported. 'validated' means
     * it exports the validated data if available, 'participant' is the original
     * participant responses */
    responseType: 'participant' | 'validatedIfAvailable';
};
