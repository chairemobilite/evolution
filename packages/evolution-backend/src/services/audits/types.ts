/*
 * Copyright 2025, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

import { ExtendedPlaceAttributes } from 'evolution-common/lib/services/baseObjects/Place';
import { ExtendedPersonAttributes } from 'evolution-common/lib/services/baseObjects/Person';
import { ExtendedHouseholdAttributes } from 'evolution-common/lib/services/baseObjects/Household';
import { ExtendedJourneyAttributes } from 'evolution-common/lib/services/baseObjects/Journey';
import { ExtendedSegmentAttributes } from 'evolution-common/lib/services/baseObjects/Segment';
import { ExtendedVisitedPlaceAttributes } from 'evolution-common/lib/services/baseObjects/VisitedPlace';
import { ExtendedTripAttributes } from 'evolution-common/lib/services/baseObjects/Trip';
import { CorrectedResponse } from 'evolution-common/lib/services/questionnaire/types';

/**
 * Parser function type for converting survey response values to proper types
 * before object validation. For example, converting 'yes'/'no' strings to boolean values.
 * Parsers modify the corrected response in place.
 */
export type SurveyObjectParserInterview<TInput> = (attributes: TInput) => void;
export type SurveyObjectParser<TInput, CorrectedResponse> = (
    attributes: TInput,
    correctedResponse: CorrectedResponse
) => void;

/**
 * Configuration for survey object parsers.
 * Each parser takes the corrected response and modifies the response data in place
 * to convert string choice values to proper types before object creation.
 */
export type SurveyObjectParsers = {
    interview?: SurveyObjectParserInterview<CorrectedResponse>;
    household?: SurveyObjectParser<ExtendedHouseholdAttributes, CorrectedResponse>;
    home?: SurveyObjectParser<ExtendedPlaceAttributes, CorrectedResponse>;
    person?: SurveyObjectParser<ExtendedPersonAttributes, CorrectedResponse>;
    journey?: SurveyObjectParser<ExtendedJourneyAttributes, CorrectedResponse>;
    visitedPlace?: SurveyObjectParser<ExtendedVisitedPlaceAttributes, CorrectedResponse>;
    trip?: SurveyObjectParser<ExtendedTripAttributes, CorrectedResponse>;
    segment?: SurveyObjectParser<ExtendedSegmentAttributes, CorrectedResponse>;
};
