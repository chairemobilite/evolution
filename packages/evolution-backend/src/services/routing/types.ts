/*
 * Copyright 2024, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { RoutingOrTransitMode } from 'chaire-lib-common/lib/config/routingModes';
import { SummaryResponse, SummarySuccessResult } from 'chaire-lib-common/lib/api/TrRouting/trRoutingApiV2';

export type RouteCalculationParameter = {
    /**
     * The origin of the route to calculate
     */
    origin: GeoJSON.Point | GeoJSON.Feature<GeoJSON.Point>;
    /**
     * The destination of the route to calculate
     */
    destination: GeoJSON.Point | GeoJSON.Feature<GeoJSON.Point>;
    /**
     * The time of the trip, in seconds since midnight
     */
    departureSecondsSinceMidnight: number;
    /**
     * The date of the trip, in the format 'YYYY-MM-DD'
     */
    departureDateString: string;
    /**
     * If `transit` mode is requested, the scenario to use for calculations
     */
    transitScenario?: string;
};

type RoutingTimeDistanceResult = {
    // The source of the routing calculation. Can be used to identify the method
    // called for this calculation
    source: string;
} & (
    | {
          status: 'no_routing_found';
      }
    | {
          status: 'error';
          error: string;
      }
    | {
          status: 'success';
          distanceM: number;
          travelTimeS: number;
      }
);

export type RoutingTimeDistanceResultByMode = {
    [mode in RoutingOrTransitMode]?: RoutingTimeDistanceResult;
};

export type SummaryResult = {
    // The source of the summary calculation. Can be used to identify the method
    // called for this calculation
    source: string;
} & (
    | {
          status: 'no_routing_found';
      }
    | {
          status: 'error';
          error: string;
      }
    | {
          status: 'success';
          nbRoutes: number;
          lines: SummarySuccessResult['result']['lines'];
      }
);
