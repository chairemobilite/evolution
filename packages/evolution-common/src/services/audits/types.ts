/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

export type Audit = {
    /**
     * Indicate the current version of a validation. If the validation changes, the version will change
     */
    version: number;
    /**
     * Whether this audit message is simply a warning or an error. Defaults to false
     */
    isWarning?: boolean;
    /**
     * The code of this validation. It should be unique per interview object.
     */
    errorCode: string;
    /**
     * A valid i18n key to translate this message. If not set, the errorCode will be used as the i18n key.
     */
    message?: string;
    /**
     * Whether to ignore this validation in the interface. Defaults to false
     */
    ignore?: boolean;
};

export type AuditForObject = Audit & {
    /**
     * The type of the object this audit applies to. Can be 'interview', 'home', 'person', 'household', etc.
     */
    objectType: string;
    /**
     * The object's UUID. If the object is a single object without a specific uuid, like the interview or household, the interview's uuid is used for this field.
     */
    objectUuid: string;
};

export type Audits = { [key: string]: Audit };
