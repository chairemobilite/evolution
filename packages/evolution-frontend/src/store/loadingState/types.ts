/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
export enum LoadingStateActionTypes {
    INCREMENT_LOADING_STATE = 'INCREMENT_LOADING_STATE',
    DECREMENT_LOADING_STATE = 'DECREMENT_LOADING_STATE',
}

export type LoadingStateAction =
    | {
          type: LoadingStateActionTypes.INCREMENT_LOADING_STATE;
          loadingState: number;
      }
    | {
          type: LoadingStateActionTypes.DECREMENT_LOADING_STATE;
          loadingState: number;
      };

export interface LoadingStateState {
    readonly loadingState?: number;
}
