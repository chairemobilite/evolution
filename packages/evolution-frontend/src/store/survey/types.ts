/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
export enum SurveyActionTypes {
    SET_INTERVIEW = 'SET_INTERVIEW',
    UPDATE_INTERVIEW = 'UPDATE_INTERVIEW'
}

export type SurveyAction =
    | {
          type: SurveyActionTypes.SET_INTERVIEW;
          // TODO Properly type the interview
          interview: any;
          interviewLoaded: boolean;
      }
    | {
          type: SurveyActionTypes.UPDATE_INTERVIEW;
          // TODO Properly type the interview
          interview: any;
          interviewLoaded: boolean;
          errors: {
              [key: string]: {
                  [key: string]: string;
              };
          };
          submitted: boolean;
      };

export interface SurveyState {
    readonly interview?: any;
    readonly interviewLoaded?: boolean;
    readonly errors?: {
        [key: string]: {
            [key: string]: string;
        };
    };
    readonly submitted?: boolean;
}
