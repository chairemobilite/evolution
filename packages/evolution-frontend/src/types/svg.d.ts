/*
 * Copyright 2025, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

// FIXME See if we need to move this declaration to the root of the project to allow other typescript files to use it
declare module '*.svg' {
    const content: any;
    export default content;
}
