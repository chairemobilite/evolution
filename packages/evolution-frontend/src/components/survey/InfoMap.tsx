/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React, { useEffect, useState } from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';
import { GoogleMap, useJsApiLoader, Marker, Polyline, Polygon, MarkerProps } from '@react-google-maps/api';
import Markdown from 'react-markdown';
import remarkGfm from 'remark-gfm';

import projectConfig from 'chaire-lib-common/lib/config/shared/project.config';
import { getCurrentGoogleMapConfig } from '../../config/googleMaps.config';
import * as surveyHelper from 'evolution-common/lib/utils/helpers';
import InputLoading from '../inputs/InputLoading';
import { UserInterviewAttributes } from 'evolution-common/lib/services/questionnaire/types';
import { CliUser } from 'chaire-lib-common/lib/services/user/userType';
import { WidgetStatus } from 'evolution-common/lib/services/questionnaire/types';
import { InfoMapWidgetConfig } from 'evolution-common/lib/services/questionnaire/types';

// TODO This is a google info map. Either it remains google and we should see if
// we can extract common code with the InputMapGoogle widget, or transform to
// use other map engine with OpenStreetMap instead.

type InfoMapProps = {
    widgetConfig: InfoMapWidgetConfig;
    widgetStatus: WidgetStatus;
    interview: UserInterviewAttributes;
    user?: CliUser;
    path: string;
};

const InfoMap: React.FC<InfoMapProps & WithTranslation> = (props: InfoMapProps & WithTranslation) => {
    // Set the google map config once, as it cannot be changed after it is
    // loaded (for language change for example). see
    // https://stackoverflow.com/questions/7065420/how-can-i-change-the-language-of-google-maps-on-the-run
    const googleMapConfig = React.useMemo(() => getCurrentGoogleMapConfig(props.i18n.language), []);
    const { isLoaded } = useJsApiLoader(googleMapConfig);

    const [map, setMap] = React.useState<google.maps.Map | null>(null);
    const [center, setCenter] = React.useState<{ lat: number; lng: number } | google.maps.LatLng>({
        lat: projectConfig.mapDefaultCenter.lat,
        lng: projectConfig.mapDefaultCenter.lon
    });

    useEffect(() => {
        const configDefaultCenter = surveyHelper.parse(
            props.widgetConfig.defaultCenter,
            props.interview,
            props.path,
            props.user
        );
        const defaultCenter = configDefaultCenter
            ? { lat: configDefaultCenter.lat, lng: configDefaultCenter.lon }
            : { lat: projectConfig.mapDefaultCenter.lat, lng: projectConfig.mapDefaultCenter.lon };
        setCenter(defaultCenter);
    }, [props.widgetConfig.defaultCenter]);

    const onMapReady = React.useCallback((map: google.maps.Map) => {
        map.panTo(center);
        setMap(map);
    }, []);

    const onUnmount = React.useCallback(() => {
        setMap(null);
    }, []);

    const onZoomChange = () => {
        if (!map) return;
        const currentZoom = (map as google.maps.Map).getZoom();
        if (currentZoom && props.widgetConfig.maxZoom && props.widgetConfig.maxZoom < currentZoom) {
            (map as google.maps.Map).setZoom(props.widgetConfig.maxZoom);
        }
    };

    if (!props.widgetStatus.isVisible) {
        return null;
    }

    if (!isLoaded) {
        return <InputLoading />;
    }

    const title =
        surveyHelper.translateString(props.widgetConfig.title, props.i18n, props.interview, props.path, props.user) ||
        '';
    const geojsons = props.widgetConfig.geojsons(props.interview, props.path, undefined);
    const gMarkers: JSX.Element[] = [];
    const gPolylines: JSX.Element[] = [];
    const gPolygons: JSX.Element[] = [];
    const points = geojsons.points?.features || [];
    const linestrings = geojsons.linestrings?.features || [];
    const polygons = geojsons.polygons?.features || [];

    const linestringColor = props.widgetConfig.linestringColor || '#0000ff';
    const linestringActiveColor = props.widgetConfig.linestringActiveColor || '#00ff00';

    const bounds = new google.maps.LatLngBounds();

    for (let i = 0, countI = points.length; i < countI; i++) {
        const point = points[i];
        const gLatLng = new google.maps.LatLng(point.geometry.coordinates[1], point.geometry.coordinates[0]);
        const markerParams: MarkerProps = {
            position: gLatLng,
            draggable: false
        };
        if (point.properties.icon) {
            markerParams.icon = {
                url: point.properties.icon.url,
                size: new google.maps.Size(point.properties.icon.size[0], point.properties.icon.size[1]),
                scaledSize: new google.maps.Size(point.properties.icon.size[0], point.properties.icon.size[1])
            };
        }
        gMarkers.push(<Marker key={`gMarker_infoMap_${props.path}__${i}`} {...markerParams} />);
        if (bounds) {
            bounds.extend({ lat: gLatLng.lat(), lng: gLatLng.lng() });
        }
    }

    for (let i = 0, countI = linestrings.length; i < countI; i++) {
        const linestring = linestrings[i];
        const polylineCoordinates = linestring.geometry.coordinates.map((coordinates: number[]) => ({
            lat: coordinates[1],
            lng: coordinates[0]
        }));

        const gArrowWhite = {
            path: google.maps.SymbolPath.FORWARD_CLOSED_ARROW,
            strokeColor: '#ffffff',
            strokeOpacity: 0.8,
            strokeWeight: 5,
            scale: 2,
            fillColor: '#ffffff',
            fillOpacity: 0.8
        };

        const gArrowBlack = {
            path: google.maps.SymbolPath.FORWARD_CLOSED_ARROW,
            strokeColor: '#000000',
            strokeOpacity: 0.4,
            strokeWeight: 7,
            scale: 2,
            fillColor: '#000000',
            fillOpacity: 0.4
        };

        const gArrow = {
            path: google.maps.SymbolPath.FORWARD_CLOSED_ARROW,
            strokeColor:
                linestring.properties && linestring.properties.active ? linestringActiveColor : linestringColor,
            strokeOpacity: 0.6,
            strokeWeight: 3,
            scale: 2,
            fillColor: linestring.properties && linestring.properties.active ? linestringActiveColor : linestringColor,
            fillOpacity: 0.6
        };

        gPolylines.push(
            <Polyline
                key={`gPolyline_infoMap_black_${props.path}__${i}`}
                path={polylineCoordinates}
                options={{
                    strokeColor: '#000000',
                    strokeWeight: 7,
                    strokeOpacity: 0.4,
                    icons: [
                        {
                            icon: gArrowBlack,
                            offset: '50%'
                        }
                    ],
                    zIndex: 300 || Math.ceil((1000.0 / i) * 1000.0)
                }}
            />
        );

        gPolylines.push(
            <Polyline
                key={`gPolyline_infoMap_white_${props.path}__${i}`}
                path={polylineCoordinates}
                options={{
                    strokeColor: '#ffffff',
                    strokeWeight: 5,
                    strokeOpacity: 0.8,
                    icons: [
                        {
                            icon: gArrowWhite,
                            offset: '50%'
                        }
                    ],
                    zIndex: 400 || Math.ceil((1000.0 / i) * 1000.0)
                }}
            />
        );

        gPolylines.push(
            <Polyline
                key={`gPolyline_infoMap_${props.path}__${i}`}
                path={polylineCoordinates}
                options={{
                    strokeColor:
                        linestring.properties && linestring.properties.active ? linestringActiveColor : linestringColor,
                    strokeWeight: 3,
                    strokeOpacity: 0.4,
                    icons: [
                        {
                            icon: gArrow,
                            offset: '50%'
                        }
                    ],
                    zIndex: 500 || Math.ceil((1000.0 / i) * 1000.0)
                }}
            />
        );
    }

    for (let i = 0, countI = polygons.length; i < countI; i++) {
        const polygon = polygons[i];
        const polygonCoordinates = polygon.geometry.coordinates[0].map((coordinates: number[]) => ({
            lat: coordinates[1],
            lng: coordinates[0]
        }));

        gPolygons.push(
            <Polygon
                key={`gPolygon_infoMap_${props.path}__${i}`}
                paths={polygonCoordinates}
                options={{
                    strokeColor: polygon.properties.strokeColor || '#0000FF',
                    strokeOpacity: polygon.properties.strokeOpacity || 0.8,
                    strokeWeight: polygon.properties.strokeWeight || 2,
                    fillColor: polygon.properties.fillColor || '#0000FF',
                    fillOpacity: polygon.properties.fillOpacity || 0.35
                }}
            />
        );
        if (
            bounds &&
            polygon.properties.minLat &&
            polygon.properties.maxLat &&
            polygon.properties.minLong &&
            polygon.properties.maxLong
        ) {
            bounds.extend({ lat: polygon.properties.minLat, lng: polygon.properties.minLong });
            bounds.extend({ lat: polygon.properties.maxLat, lng: polygon.properties.maxLong });
            bounds.extend({ lat: polygon.properties.minLat, lng: polygon.properties.maxLong });
            bounds.extend({ lat: polygon.properties.maxLat, lng: polygon.properties.minLong });
        }
    }

    if (map && bounds) {
        map.fitBounds(bounds);
    }

    return (
        <div className="survey-info-map__map-container">
            <Markdown remarkPlugins={[[remarkGfm, { singleTilde: false }]]} className="infoMap-title">
                {title}
            </Markdown>
            <GoogleMap
                id="info-map"
                mapContainerStyle={{
                    boxSizing: 'border-box',
                    position: 'relative',
                    width: '100%',
                    height: '400px',
                    border: '1px solid rgba(0,0,0,0.2'
                }}
                center={center}
                onLoad={onMapReady}
                onUnmount={onUnmount}
                onZoomChanged={onZoomChange}
                clickableIcons={false}
                zoom={props.widgetConfig.defaultZoom || 10}
            >
                {gMarkers}
                {gPolylines}
                {gPolygons}
            </GoogleMap>
        </div>
    );
};

export default withTranslation()(InfoMap);
