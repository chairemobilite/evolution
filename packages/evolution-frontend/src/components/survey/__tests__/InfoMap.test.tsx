/*
 * Copyright 2024, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import _cloneDeep from 'lodash/cloneDeep';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import each from 'jest-each';
import { render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
expect.extend(toHaveNoViolations);

import { interviewAttributes } from '../../inputs/__tests__/interviewData.test';
import InfoMap from '../InfoMap';
import { WidgetStatus } from 'evolution-common/lib/services/questionnaire/types';

// Mock react-markdown and remark-gfm as they use syntax not supported by jest
jest.mock('react-markdown', () => 'Markdown');
jest.mock('remark-gfm', () => 'remark-gfm');

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
};

const commonWidgetConfig = {
    type: 'infoMap' as const,
    title: 'title',
    geojsons: {}
};

const defaultWidgetStatus: WidgetStatus = {
    path: 'foo',
    isVisible: true,
    modalIsOpen: false,
    isDisabled: false,
    isCollapsed: false,
    isEmpty: false,
    isCustomEmpty: false,
    isValid: true,
    isResponded: true,
    isCustomResponded: false,
    errorMessage: undefined,
    currentUpdateKey: 1,
    value: null
};

beforeEach(() => {
    jest.clearAllMocks();
});

each([
    ['Default values', commonWidgetConfig],
    ['All values set', {
        ...commonWidgetConfig,
        path: 'home.region',
        conditional: jest.fn(),
        defaultCenter: { lat: 0, lon: 0 },
        maxZoom: 18,
        defaultZoom: 16,
        linestringColor: 'blue',
        linestringActiveColor: 'red'
    }],
]).describe('Button widget: %s', (_widget, widgetConfig) => {

    test('Render widget', () => {

        const wrapper = TestRenderer.create(
            <InfoMap
                path='home.region'
                widgetConfig={widgetConfig}
                interview={interviewAttributes}
                user={userAttributes}
                widgetStatus={defaultWidgetStatus}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Widget accessibility', async () => {
        const { container } = render(
            <InfoMap
                path='home.region'
                widgetConfig={widgetConfig}
                interview={interviewAttributes}
                user={userAttributes}
                widgetStatus={defaultWidgetStatus}
            />
        );
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
});
