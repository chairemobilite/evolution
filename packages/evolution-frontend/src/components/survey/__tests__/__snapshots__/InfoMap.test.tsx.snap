// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button widget: All values set Render widget 1`] = `
<div
  className="loader"
>
  <p>
    Loading
    ...
  </p>
</div>
`;

exports[`Button widget: Default values Render widget 1`] = `
<div
  className="loader"
>
  <p>
    Loading
    ...
  </p>
</div>
`;
